### ISSUE TITLE
Please add supports to skorch

### ISSUE THREAD

**Issue by UTUnex** [optuna/optuna#1654]
Skorch is a wrapper of pytorch which makes it have a sklearn API format for easy use. Currently, it's a pity that I didn't find support of optuna for skorch. You know that there are still many users who just transfer from sklearn to neural network and still want to keep our familiar sklearn-style:)

Recently I have been searching for the ways to apply the optuna to my skorch program but I didn't find any solutions and examples and I have to turn to other tools which is very disappointing...

Anyway, please consider my request:)

**Comment by HideakiImamura** [comment]
Thank you for the interesting suggestion! Do you have any skorch's code in which you want to use Optuna? It will help us how we integrate the skorch with Optuna.

**Comment by UTUnex** [comment]
> Thank you for the interesting suggestion! Do you have any skorch's code in which you want to use Optuna? It will help us how we integrate the skorch with Optuna.

Thank you for your reply. The following is the code from skorch documenation:


# %%
import torch
from torch import nn
import torch.nn.functional as F

torch.manual_seed(0);

# %%
import numpy as np
from sklearn.datasets import make_classification

# %%
X, y = make_classification(1000, 20, n_informative=10, random_state=0)
X, y = X.astype(np.float32), y.astype(np.int64)

# %%
X.shape, y.shape, y.mean()

# %%
class ClassifierModule(nn.Module):
    def __init__(
            self,
            num_units=10,
            nonlin=F.relu,
            dropout=0.5,
    ):
        super(ClassifierModule, self).__init__()
        self.num_units = num_units
        self.nonlin = nonlin
        self.dropout = dropout

        self.dense0 = nn.Linear(20, num_units)
        self.nonlin = nonlin
        self.dropout = nn.Dropout(dropout)
        self.dense1 = nn.Linear(num_units, 10)
        self.output = nn.Linear(10, 2)

    def forward(self, X, **kwargs):
        X = self.nonlin(self.dense0(X))
        X = self.dropout(X)
        X = F.relu(self.dense1(X))
        X = F.softmax(self.output(X), dim=-1)
        return X

# %%
from skorch import NeuralNetClassifier

# %%
net = NeuralNetClassifier(
    ClassifierModule,
    max_epochs=20,
    lr=0.1,
#     device='cuda',  # uncomment this to train with CUDA
)

# %%
net.fit(X, y)

# %%
y_pred = net.predict(X[:5])
y_pred


# %%
y_proba = net.predict_proba(X[:5])
y_proba

# %%


>>[1]What I hope to do is to tune the **num_units**(number of neurons in hidden layer), the **dropout**, the nonlin(**activation function**)

>>[2]The expected tuner API format should be like : optunasearchcv(**kwargs) which is better close to the sklearn GridsearchCV format

>>[3]The expected searching algorithms: Bayesian Optimization based algorithms like TPE; BOHB;
>>[4]Please consider single GPU usage case; multiple GPU usage case;
>>[5]Please consider the cross validation control (whether or not) because as is known for many cases CV is just computational expensive for  NN model selection
>>[6]The most expected thing is the example, for many common users like me, a good example is always a good start 
>>[7]Clear documentation is also expected
>>[8]The integration with some monitoring framework like tensorboard or ML flow

Anyway, I hope you can consider my expectation and make them all implemented in optuna.

Thank you.

**Comment by HideakiImamura** [comment]
Thank you for the detailed explanations. PR #1668 is related to this issue.

**Comment by BenjaminBossan** [comment]
Just FYI, I'm currently adding an [example](https://github.com/skorch-dev/skorch/pull/724) of using skorch and optuna to the skorch repo.

I started that without knowledge of [SkorchPruningCallback](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.SkorchPruningCallback.html) but came up with essentially the same implementation. That gives me confidence that this is the correct approach :)

My example [notebook](https://github.com/skorch-dev/skorch/blob/feature/optuna-example-718/notebooks/optuna-example.ipynb) differs a bit from the [example](https://github.com/optuna/optuna/blob/master/examples/skorch_simple.py) in the optuna repo. Most notably, I moved the `trial.suggest*` code out from the module's `__init__` method. I think this is better because this way, the module stands on its own and can be used without a `trial` instance.

Since I don't know a lot about optuna, feel free to share your feedback on how to improve the notebook!
